cinn/runtime/cuda/cuda_util.cc

```diff
@@ -30,43 +32,90 @@ DEFINE_bool(cinn_cudnn_deterministic,
             "operator. The autotuning algorithm may be non-deterministic. If "
             "true, the algorithm is deterministic.");
 
+DEFINE_int32(cinn_cudnn_max_ws, 0, "The limit of cudnn workspace size. 0 means that no limit exist!");
+
 namespace cinn {
 namespace runtime {
 namespace cuda {
 
-SerialData::SerialData() {}
+CublasHandle::CublasHandle() { cublasCreate(&cublas); }
+
+CublasHandle::~CublasHandle() { cublasDestroy(cublas); }
 
-SerialData::~SerialData() {}
+class BufferWrapper;
+class CudnnHelper {
+ public:
+  static CudnnHelper &Instance() {
+    static CudnnHelper cudnn_helper;
+    return cudnn_helper;
+  }
 
-CudnnHandle::CudnnHandle() {
-  CUDNN_CALL(cudnnCreate(&cudnn));
-  size_      = 0;
-  work_space = nullptr;
-}
+  cudnnHandle_t GetCudnnHandle() { return handle_t; }
 
-CudnnHandle::~CudnnHandle() {
-  CUDNN_CALL(cudnnDestroy(cudnn));
-  if (size_ > 0) {
-    CUDA_CALL(cudaFree(work_space));
+  void RegisterAlgo(const std::string &key, const int value) {
+    static std::mutex local_mtx;
+    std::lock_guard<std::mutex> lock(local_mtx);
+    cudnn_algo_map[key] = value;
   }
-}
 
-CublasHandle::CublasHandle() { cublasCreate(&cublas); }
+  int GetAlgo(const std::string &key) {
+    if (cudnn_algo_map.count(key) == 0) {
+      return -1;
+    } else {
+      return cudnn_algo_map[key];
+    }
+  }
 
-CublasHandle::~CublasHandle() { cublasDestroy(cublas); }
+  BufferWrapper GetWorkSpace(size_t size);
 
-float *CudnnHandle::GetWorkSpace(size_t size) {
-  if (size_ >= size) {
-    return work_space;
-  } else {
-    if (size_ > 0) {
-      CUDA_CALL(cudaFree(work_space));
+  void ReleaseWS(void *ptr, size_t size) {
+    std::lock_guard<std::mutex> lock(mtx);
+    cudnn_ws_map.push(std::pair<void *, size_t>(ptr, size));
```

Comment:
Please consider rename `cudnn_ws_map`, this doesn't give the meaning that it temporary stores the released workspace `std::pair<void *, size_t>`

Reply:
done,thanks

Reply:
cudnn_ws_map->cudnn_ws_queue